### PR TITLE
Add lambertian renderer

### PR DESCRIPTION
**Pull request by terrancewang**

### CODE REVIEW

File: nerfstudio/model_components/renderers.py [hunk 1]
```diff
@@ -332,3 +332,32 @@ def forward(
...
+
+class LambertianShadingRenderer(nn.Module):
+    """Calculate Lambertian shading."""
+
+    @classmethod
+    def forward(
+        cls,
+        rgb: TensorType["bs":..., "num_samples", 3],
+        normals: TensorType["bs":..., "num_samples", 3],
+        light_direction: TensorType["bs", 3],
```
Comment (tancik) [inline nerfstudio/model_components/renderers.py:345]: should this be `["bs"..., 3]`

File: nerfstudio/model_components/renderers.py [hunk 2]
```diff
@@ -332,3 +332,32 @@ def forward(
...
+        Args:
+            rgb: Accumulated rgb along a ray.
+            normals: Accumulated normals along a ray.
+            light_direction: Direction of light source.
+            ratio: ambient lighting (1.0) vs. Lambertian shading (0.0) ratio
+
+        Returns:
+            Textureless Lambertian shading, Lambertian shading
+        """
+        lambertian = ratio + (1 - ratio) * (normals.detach() @ light_direction).clamp(min=0)
```
Comment (tancik) [inline nerfstudio/model_components/renderers.py:359]: Do we always want to detach normals? Maybe we should add an argument that allows the user to choose.

File: nerfstudio/model_components/renderers.py [hunk 3]
```diff
@@ -332,3 +332,32 @@ def forward(
...
+class LambertianShadingRenderer(nn.Module):
+    """Calculate Lambertian shading."""
+
+    @classmethod
+    def forward(
+        cls,
+        rgb: TensorType["bs":..., "num_samples", 3],
+        normals: TensorType["bs":..., "num_samples", 3],
+        light_direction: TensorType["bs", 3],
+        ratio: float = 1.0,
```
Comment (tancik) [inline nerfstudio/model_components/renderers.py:346]: Seems odd for the default to be non-lambertian. Does something like .5 make more sense?

**Review by tancik — COMMENTED** [review]
Thanks for adding.

File: nerfstudio/model_components/renderers.py (hunk 3 above)
Comment (tancik) [reply nerfstudio/model_components/renderers.py:346]: Actually thinking a bit more, I think `ratio` could use a different name like `shading_weight` where when `shading_weight = 1` the render gives you something that is pure lambertian (0 gives no shading). The default can then be 1.0.

Comment (terrancewang) [reply nerfstudio/model_components/renderers.py:346]: yeah that name is more clear

**Review by tancik — APPROVED** [review]
LGTM

File: nerfstudio/model_components/renderers.py (hunk 2 above)
Comment (mcallisterdavid) [reply nerfstudio/model_components/renderers.py]: I didn’t see this earlier, but the reason for detaching the normals was that they get supervised by your SDS loss in Dreamfusion if you don’t detach them. The optimization then makes all your normals the same direction which basically turns off lambertian shading. Maybe it’s best to just turn on “detached normals” or something for the Dreamfusion implementation

Comment (terrancewang) [reply nerfstudio/model_components/renderers.py]: do you remember if you guys ever just tried using the normals network that stable dreamfusion used? I think they were using the sds loss to supervise normals.
